Polite nrepl server

PEZ opened the pull request:
## What has changed?

Actually using the nREPL `describe` results to not spam the server with ops it doesn't handle anyway. Now the ops calls resolve to (or return, depending) `undefined` if they are not supported.

In some places I have added checks for this undefined. Not sure why that wasn't needed before, so I am pretty sure I have missed some places. But at least now I don't see any errors thrown in the console and things seem to work. Actually, they work better than before. E.g. Babashka info lookups were not succeeding because a `list-ns` call is not supported.


Fixes
* #1969

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [ ] Added to or updated docs in this branch, if appropriate
- [x] Tests
  - [x] Tested the particular change
  - [x] Figured if the change might have some side effects and tested those as well.
  - [x] Smoke tested the extension as such.
  - [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
  - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
  - [ ] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

<!-- This is a nice book to read about the power of checklists: https://www.samuelthomasdavies.com/book-summaries/health-fitness/the-checklist-manifesto/ -->

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik

PEZ committed:
Make nrepl msg functions check if op is supported

Don't send message if not, if log is enabled, log that it is not sent

PEZ committed:
Fall back on cljc session if cljs one is missing

PEZ committed:
Check 'info' available via describe

PEZ committed:
ns-list might not be available

PEZ committed:
clojuredocs op might not be available

PEZ committed:
Sometimes the promise resolves with undefined...

PEZ committed:
Don't resolve non-promises

PEZ committed:
Add not-sending nrepl log

PEZ committed:
Don't send info op for defnitions if not supported

PEZ committed:
Add test-data joyride script

PEZ committed:
Don't use info if not supported

PEZ committed:
Add joyride scripts in deps.edn paths

PEZ committed:
Add comment

PEZ committed:
Don't access list if undefined

PEZ committed:
clojure-lsp seems to have camelCased this

PEZ committed:
Wait a sec with initialize the debugger, to allow describe some time

PEZ commented:
Noticing that the testRunner now runs namespace tests twice if the tests lives in a `-test` namespace and you run them while the non-test namespace is the active file. I don't understand why this wasn't happening before, though. The `test-var-query` calls look different, which is also strange.

PEZ committed:
Load test namespace, don't also run its tests

PEZ commented:
Still not sure what was going on, but I think we were compensating for some bug by running tests for both the ns and for it's `<ns>-test` namespace. Now `run namespace tests` runs the tests for the current ns, loading `<ns>-test` first.

PEZ committed:
Change fallback strategy for stackrace call

bpringe commented on src/connector.ts:
```diff
@@ -634,21 +632,28 @@ export async function connect(
+      if (!nClient.session.supports('info')) {
+        void vscode.window
+          .showWarningMessage(
+            'The nREPL server does not support cider-nrepl `info` op, which indicates troubles ahead. You need to start the REPL with cider-nrepl dependencies met',
```
```suggestion
            'The nREPL server does not support cider-nrepl `info` op, which indicates troubles ahead. You need to start the REPL with cider-nrepl dependencies met.',
```

bpringe commented on src/namespace.ts:
```diff
@@ -75,7 +75,7 @@ export async function createNamespaceFromDocumentIfNotExists(doc) {
       if (client) {
         const nsList = await client.listNamespaces([]);
-        if (nsList['ns-list'] && nsList['ns-list'].includes(ns)) {
+        if (nsList && nsList['ns-list'] && nsList['ns-list'].includes(ns)) {
```
I think the above two lines could be a bit more simplified with optional chaining (I saw another case where this could be used in this PR too).

```typescript
        const nsList = await client.listNamespaces([])?.nsList;
        if (nsList?.includes(ns)) {
```

bpringe replied:
Though this is arguably more clear, I suppose - just a bit less code.

PEZ replied:
It looked like that at first, but then I changed it for reasons I don't remember. =)

bpringe commented on src/nrepl/index.ts:
```diff
@@ -393,21 +431,32 @@ export class NReplSession {
       session: this.sessionId,
-    });
+    };
+    if (this.supports('stdin')) {
```
I think it would be better to check for support inside of the `write` function, rather than having to check in every function that calls `write`. Then this if check + log statement (if the op isn't supported), is only in one place.

bpringe replied:
Maybe make `write` return `true` if the op is supported and was called or `undefined` if it's not and was not, then return the result of `write` (or pass the result to `resolve`) in these op calling functions.

PEZ replied:
I had it in the write function at first. But then it looked wrong to do a bunch of maybe-writes, and I also had some problems with it. The functions need to resolve with the result of the op, so I still had logic in the functions checking the result of the write...

But I reverted to this way without investigating the options to press on with the centralized approach. I guess if we pass on the resultResolver to write, it can handle this and we would have even less duplicated code. I'll have a look, but maybe we need to write it up on the todo list.

bpringe replied:
Ah maybe it's not as simple as I was thinking. Also I see in some cases logic specific to the function occurs conditionally based on support for the op before write is called. There's probably a way to refactor it so put the support check in `write` but yeah... a todo item works for now.

PEZ replied:
Yeah, I'm pretty sure it can be done, but all my attempts get very messy so far. Someone with better command over JavaScript than I have might see what should be done.

PEZ committed:
Update src/connector.ts

Co-authored-by: Brandon Ringe <12722744+bpringe@users.noreply.github.com>

PEZ committed:
Log error when interrupt not supported

Actually want to prevent the message that the interrupt happened,
but I fail to do so.

PEZ committed:
Only register message handler if op is supported

PEZ committed:
The format for clojure-lsp seeAslos has changed

PEZ committed:
Make links out of seeAlsos

PEZ committed:
Make links out of seeAlsos

Fixes #1972

PEZ committed:
Merge branch '1969-polite-nrepl-server' of github.com:BetterThanTomorrow/calva into 1969-polite-nrepl-server

PEZ committed:
Keep the nrepl log when toggling it logging off

PEZ committed:
Merge branch 'dev' into 1969-polite-nrepl-server

PEZ committed:
Wait for both session and descibe before resolving client

PEZ committed:
Don't strip clojuredocs paths of `clojure.core/`

PEZ committed:
Update Changelog

Fixes #1969
Fixes #1972

PEZ committed:
Add changelog section for v2.0.320 [skip ci]

PEZ commented:
I need to update the publish script to stop me from releasing from a feature branch...
